### PR TITLE
feat(#287): the set of tests-to-run is provided to TestExecutionPlanner implementations

### LINKED ISSUE #287: Provide set of tests-to-run to TestExecutionPlanner implementations
##### Issue Overview

Provide set of tests-to-run to TestExecutionPlanner implementations.
There are two benefits of it
* we can optimize the execution of the planners by minimizing of the scope (eg. calculating the graph for the test classes)
* make feasible implementation of new strategies that need to retrieve some concrete information from the test classes (eg. annotations - see: #283)

##### Expected Behaviour

Replace the method 
```java
Collection<TestSelection> getTests();
```
by two methods:
```java
Collection<TestSelection> selectTestsFromNames(Iterable<String> testsToRun);

Collection<TestSelection> selectTestsFromClasses(Iterable<Class<?>> testsToRun);
```

### PR DESCRIPTION
**Pull request by MatousJobanek**
#### Short description of what this resolves:

The implementations of the strategies didn't have the information about the set of tests that were selected to be executed. In this PR I'm introducing the changes in `TestExecutionPlanner` API that will provide this kind of information.

#### Changes proposed in this pull request:

- in `TestExecutionPlanner`, instead of one method `getTests()` there are two methods:
```java
Collection<TestSelection> selectTestsFromNames(Iterable<String> testsToRun);
Collection<TestSelection> selectTestsFromClasses(Iterable<Class<?>> testsToRun);
```

Fixes #287

### CODE REVIEW

File: core/src/main/java/org/arquillian/smart/testing/impl/TestStrategyApplierImpl.java [hunk 1]
```diff
@@ -35,16 +36,28 @@
...
+    private Collection<TestSelection> selectTestsFromClasses(TestExecutionPlanner testExecutionPlanner,
+        Iterable<Class<?>> testsToRun) {
+        return testExecutionPlanner.selectTestsFromClasses(testsToRun);
     }
 
     private <TESTCLASS> Set<TestSelection> apply(Iterable<TESTCLASS> testsToRun,
-        Function<TESTCLASS, String> mapperToName) {
-        final Set<TestSelection> selectedTests = selectTests(configuration);
+        Function<TESTCLASS, String> mapperToName,
+        BiFunction<TestExecutionPlanner, Iterable<TESTCLASS>, Collection<TestSelection>> selectionFunction) {
```
Comment (bartoszmajsak) [inline core/src/main/java/org/arquillian/smart/testing/impl/TestStrategyApplierImpl.java]: I am not sure if using `BiFunction` here is not making it hard to follow (plus `<TESTCLASS>` "generic"). I have a feeling that this is a smell that the code here became overcomplicated (if it wasn't already).

Comment (bartoszmajsak) [reply core/src/main/java/org/arquillian/smart/testing/impl/TestStrategyApplierImpl.java]: We have a mix of high-level, general purpose concepts such as `apply` and then within it `selectTests`. Not sure how we can make it better, but maybe a good start would be to separate `selectTests` logic from applying a strategy (so to call them one after the other instead of one from within another)?

File: docs/registering-strategies.adoc [hunk 2]
```diff
@@ -33,9 +33,17 @@ include::../strategies/failed/src/main/resources/META-INF/services/org.arquillia
...
-The important method there is `getTests()` which returns a collection of the tests that must be executed.
-You also need to set there the strategy name.
+`TestExecutionPlanner` is the implementation of the strategy.
+Apart from the method `getName()` that returns a name of the strategy there are two important methods:
+
+[source, java]
+----
+selectTestsFromNames(Iterable<String> testsToRun)
+selectTestsFromClasses(Iterable<Class<?>> testsToRun)
+----
```
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply docs/registering-strategies.adoc]: Sure, great idea

File: core/src/main/java/org/arquillian/smart/testing/impl/TestStrategyApplierImpl.java (hunk 1 above)
Comment (MatousJobanek) [reply core/src/main/java/org/arquillian/smart/testing/impl/TestStrategyApplierImpl.java]: Yeah, I do agree that it is overcomplicated - I wanted to avoid copy-pasting and code duplication.
I tried to refactor it, but I'm not sure if I made it better - I introduced a new abstraction - `TestSelector` that has two implementations (based on the class info type - classes/names). It does the whole magic and from the applier you call only `selectTests` or `orderTests`

Comment (bartoszmajsak) [reply core/src/main/java/org/arquillian/smart/testing/impl/TestStrategyApplierImpl.java]: Way nicer :) Let's see if we can further improve it going forward, but this is good enough! Thank you!

**Comment by bartoszmajsak** [comment]
borat-very-nice
